acvm/src/pwg/block.rs

```diff
@@ -67,69 +71,66 @@ impl BlockSolver {
+            let index = expression_to_const(&index_expr).ok_or_else(|| {
+                OpcodeResolutionError::OpcodeNotSolvable(OpcodeNotSolvable::MissingAssignment(
+                    index_expr.any_witness().unwrap().0,
+                ))
+            })?;
+            let index = index.try_to_u64().unwrap() as u32;
+            let value = ArithmeticSolver::evaluate(&block_op.value, initial_witness);
+            let value_witness = value.any_witness();
+            if operation.is_zero() {
+                let value = ArithmeticSolver::evaluate(&block_op.value, initial_witness);
+                if value.is_const() {
+                    self.insert_value(index, value.q_c)?;
+                } else if value.is_linear() {
+                    match ArithmeticSolver::solve_fan_in_term(&value, initial_witness) {
+                        GateStatus::GateUnsolvable => {
+                            return Err(OpcodeResolutionError::OpcodeNotSolvable(
+                                OpcodeNotSolvable::MissingAssignment(value_witness.unwrap().0),
+                            ))
+                        }
+                        GateStatus::GateSolvable(sum, (coef, w)) => {
+                            let map_value = self.get_value(index).ok_or(
+                                OpcodeResolutionError::OpcodeNotSolvable(
+                                    OpcodeNotSolvable::MissingAssignment(w.0),
+                                ),
+                            )?;
+                            insert_witness(
+                                w,
+                                (map_value - sum - value.q_c) / coef,
+                                initial_witness,
+                            )?;
                         }
-                    } else {
-                        let value =
-                            ArithmeticSolver::evaluate(&trace[op_idx].value, initial_witness);
-                        if value.is_const() {
-                            self.insert_value(index, value.q_c)?;
-                        } else {
-                            break;
+                        GateStatus::GateSatisfied(sum) => {
+                            self.insert_value(index, sum + value.q_c)?;
                         }
                     }
                 } else {
-                    unknown = index_expr.get_witness();
-                    break;
+                    return Err(OpcodeResolutionError::OpcodeNotSolvable(
+                        OpcodeNotSolvable::MissingAssignment(value_witness.unwrap().0),
+                    ));
                 }
+            } else if value.is_const() {
+                self.insert_value(index, value.q_c)?;
             } else {
-                unknown = op_expr.get_witness();
-                break;
+                return Err(OpcodeResolutionError::OpcodeNotSolvable(
+                    OpcodeNotSolvable::MissingAssignment(value_witness.unwrap().0),
```

Comment:
We can factor out the repeated errors into a function:
```rs
let missing_assignment = |witness| 
    OpcodeResolutionError::OpcodeNotSolvable(OpcodeNotSolvable::MissingAssignment(witness));
```

Reply:
Done